### ISSUE TITLE
Having `async void` method is a code smell which leads to unexpected behaviour

### ISSUE THREAD

**Issue by Robot1que** [PrismLibrary/Prism#724]
I've left a comment about having `async void Execute(object parameter)` method and that it swallows exception when you run it syncroniously:
https://github.com/PrismLibrary/Prism/commit/8aeac723220bf32394648ff13a46669235aff77d

I believe it's a bug.

**Comment by brianlagunas** [comment]
This is actually a good catch.  How about a PR?

**Comment by Robot1que** [comment]
Sorry, but could you explain what "PR" means?

**Comment by brianlagunas** [comment]
Pull Request.  It means you fix it and submit your changes so everyone can benefit.

**Comment by brianlagunas** [comment]
Also, I'm not sure why this was added to DelegateCommand in the first place.  I think there should have been a DelegateCommandAsync too, instead of trying to get the existing DelegateCommand to act in an async manner.   I'll see if I can ask Microsoft about their thinking when they did this.

**Comment by Robot1que** [comment]
Oh, yeah, sure. I'll try to make change tomorrow.
But yeah, I would rather create another class for `async` commands.

**Comment by bartlannoeye** [comment]
Been a while since I ran my async checker plugin indeed, great catch.

**Comment by dvorn** [comment]
I think this is by design.  DelegateCommand is an asynchronous command, even if it was defined via a synchronous Action:
```csharp
       _executeMethod = (arg) => { executeMethod(arg); return Task.Delay(0); };
```
Correct method of testing the command is instead of
```csharp
[Test]
public void Test_should_fail_because_of_thrown_exception()
{
    new DelegateCommand(() => { throw new Exception(); }).Execute();
}
```
to use 
```csharp
[Test]
public async Task Test_should_fail_because_of_thrown_exception()
{
    var command = new DelegateCommand(() => { throw new Exception(); })
    await command.Execute();
}
```The method `ICommand.Execute` has nothing to do with swallowing exceptions in this case.  ICommand.Execute is calling DelegateCommand.Execute and not conversely.

`async void ICommand.Execute` is called by the WPF and is in fact an event handler, and it is OK to have `async void` event handlers.

Thus, if you want to test a command, test the `Task DelegateCommand.Execute` method.

Note also that if a command is defined via a synchronous Action, the Action is executed synchronously before ICommand.Execute returns.  

Thus, having the asynchronous DelegateCommand provides the same behavior as a fully synchronous variant (in the previous versions of Prism) except for the slight difference in exception handling.

I think that was the design decision to not have two separate DelegateCommands - sync and async.

**Comment by dvorn** [comment]
Note also that if you define
```csharp
void ICommand.Execute(object parameter)
{
    Execute(parameter).Wait();
}
```
in most cases it will deadlock with truly asynchronous commands.  For instance, the command will need to retrieve some data asynchronously and then update some GUI elements.   Updating the GUI will block because the GUI thread is in Wait() waiting the command to complete.

**Comment by brianlagunas** [comment]
@dvorn great insight!  We will consider it by design.

**Comment by Robot1que** [comment]
@dvorn, in our design view models return instances of `ICommand` rather then `DelegateCommand` because we want to avoid changing interface of view models if we decide to change MVVM framework. If we make execute method `async await` then we are aware of it and know what to expect. On the other hand building it into framework without warning about it in a class description might confuse users of this library as it happened to me. It's not OK when users of a library have to read source code of a that library in order to understand it's behavior.

Also as far as I can see there is no way I can pass `Func<Task>` as `executeMethod` constructor parameter because that constructor is `protected`. Which means I cannot compile the following code:
```
[Test]
public async Task This_test_should_fail_because_of_thrown_exception()
{
    var command = new DelegateCommand(this.Method);
    await command.Execute();
}

private async Task Method()
{
    await Task.Delay(0);
    throw new Exception();
}
```

This means that there is no benefit in using `Func<Task>` in internal implementation of `DelegateCommandBase`. If I need to execute my code asynchronously I just pass `async void` method as `executeMethod` constructor argument just like we usually do with event handlers.

Or maybe the problem is that `protected DelegateCommand(Func<Task> executeMethod)` is protected?

**Comment by brianlagunas** [comment]
To use the async version of DelegateCommand use DelegateCommand.FromAsyncHandler

**Comment by dbeavon** [comment]
I'm experiencing exceptions being swallowed as well.  

Can someone tell us, in layman's terms, how the DelegateCommand's Execute() method syntax is supposed to be used so that exceptions aren't swallowed and my app doesn't continue running after a fatal exception (like a walking dead).

I've unwittingly used the DelegateCommand's Execute().

But the odd thing is that, whereas the DelegateCommand's Execute() exceptions get swallowed, I can simply cast to ICommand and call Execute(null) on the interface and then the exceptions cause the app to crash as expected.

I'm not yet very familiar with the async/await service in C#, but in this context the stuff seems to be causing quite a bit of confusion for simple synchronous scenarios.  I agree that I shouldn't have to be digging into the source code of DelegateCommand as have been today for a couple hours.



As a side I should add that even the cast to ICommand and the call to ICommand.Execute(null) does not crash the app as directly as I would have liked.  The app crashes in a two-step process, first the command failure is trapped via this method here:

 **System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task task)**

And subsequently crashes via the dispatcher like so:

**System.Runtime.CompilerServices.AsyncMethodBuilderCore.ThrowAsync.AnonymousMethod__6_0(object state)**

**Comment by dbeavon** [comment]
To recap, both the call to DelegateCommand.Execute() and ICommand.Execute(null) walk out from the method call and continue on their way in the current callstack from which they were used.  

The only difference is that if ICommand.Execute(null) is used, the dispatcher comes right back and hits me with the following which causes the app to finally crash.

**System.Runtime.CompilerServices.AsyncMethodBuilderCore.ThrowAsync.AnonymousMethod__6_0(object state)**

Given that **both** of the "Execute()" method calls continue on their way after failure, is there a way to detect that a fatal error had occurred and short-circuit any other code that would otherwise follow?  No point in doing more work if the dispatcher is about to crash the app.

**Comment by brianlagunas** [comment]
This issue is not unique to DelegateCommand.  You should add exception handling to your code.

**Comment by dvorn** [comment]
@dbeavon The DelegateCommand.Execute() returns a Task object which contains the status and result of the operation, in particular the thrown exceptions.  If you want your exceptions to propagate, you need to "observe" this task - e.g., `await DelegateCommand.Execute()`.  Otherwise, the exceptions is kept in the Task until it is garbage-collected (and even then it depends...).

**Comment by Robot1que** [comment]
I think what @dvorn suggested is to put `await` every time you need to call `DelegateCommand.Execute()` just like he did in his example:
```
public async Task SomeMethod()
{
    var command = new DelegateCommand(() => { throw new Exception(); });
    await command.Execute();
}
```

Which means that:
- we have to refactor our synchronous methods into asynchronous by adding `async` keyword (note that method which calls `SomeMethod()` should also be refactored this way),
- we should not operate with `ICommand` interface but rather with `DelegateCommand` class instances so we can call `public Task Execute()` method.

**Comment by dvorn** [comment]
Unfortunately this won't help when WPF calls your command via some binding (and this is the main use case).

**Comment by dbeavon** [comment]
For now I'll stick with casting to ICommand and running its Execute(null).  Seems to crash as desired.  I'm assuming that my WPF bindings are working properly because they are based on the ICommand behavior as well.

@brianlagunas  thanks for the very obvious pointer.  The point is that I have already caught and dealt with all the exceptions I care about.  The remainder of the exceptions are totally unexpected and above my pay-grade;  so those are the ones that need to crash the app.  I've learned to never trust an app that shows you a fatal error message and then doesn't crash.  The kinds of behaviors that follow from one of these apps are truly amazing.

Of course WPF swallows exceptions in some other contexts as well.  But it doesn't seem right in DelegateCommand.Execute().  Maybe I'll add an extension method called ExecuteWithoutSwallowing.  ;)

**Comment by brianlagunas** [comment]
@dbeavon An exception should never crash an app.  All exceptions should be handled and logged accordingly.  Working with an ICommand, be it DelegateCommand, RelayCommand, or your own custom command, is no different than working with an even handler in code-behind.  Event handlers are void, and when you run async/await code in an event handler, you must have proper exception handling.

**Comment by brianlagunas** [comment]
Let's address this first guideline of async/await, you should prefer async Task to async void... The exception to this guideline is asynchronous event handlers, which must return void. This exception includes methods that are logically event handlers even if they’re not literally event handlers (for example, ICommand.Execute implementations).

**Comment by brianlagunas** [comment]
I'll add that I think the original Prism team made a poor design decision trying to stuff async/await support into the existing DelegateCommand.  I would much prefer to see a separate AsyncDelegateCommand that is responsible for async/await scenarios.  Or even possibly adding a new IAsyncCommand interface to DelegateCommand.

**Comment by dbeavon** [comment]
Logging exceptions and/or showing them to the user is one thing.  Pushing forward and trying to keep your application running in the context of a totally unforeseeable and untested exception is another matter.

@brianlagunas  "An exception should never crash an app".  Seriously?  What would your exception handling look like if you received an System.OutOfMemoryException?

Of course I'd catch and handle all the possible exceptions I'm aware of, but its hubris to think any app can properly recover from totally unforeseeable exceptions.  In other words, if you haven't tested for a given exception, your app has a low likelihood of working properly when it is encountered.  That's why it needs to crash.

**Comment by brianlagunas** [comment]
@dbeavon I'm not here to discuss how to manage exception handling in apps, but rather to determine how we could improve Prism to address this new concern with the introduction of async/await.  I made some suggestions above.  Did you give those any thought?

**Comment by dbeavon** [comment]
My main thought is probably to avoid DelegateCommand.Execute() as odd as that sounds, and only use it for async scenarios.  That method seems to be designed for the 20% usage scenarios and doesn't behave properly in the 80%.

I'll stick with casting to ICommand and running its Execute(null).  That seems to be how WPF interacts with these DelegateCommands and the Prism API has little choice but to behave in a fairly reasonable way.

I expect this topic will come up from time to time when others are building fairly standard line-of-business applications and discover that the apps aren't crashing when they should.  You would be amazed how end-users will send you a screenshot of some bizarre thing that is happening in the app, without mentioning another - preceeding and more significant - error message (which they think isn't relevant; but turns out to be the root cause for some form of latent viewmodel corruption).

**Comment by dvorn** [comment]
@brianlagunas I agree that having a separate AsyncDelegateCommand might be a better choice.  And factory method FromAsyncHandler seems strange at a very least.  (Although IAsyncCommand seems pointless since WPF does not know about it.)

But what about backward compatibility?

**Comment by brianlagunas** [comment]
Honestly, I'm not sure how to go about this without breaking someone.  I don't like the current design of DelegateCommand, but it is one of the most used classes in Prism.  I will have to really think about how we can modify DelegateCommand to do what we want, and also the possibility of introducing a new AsyncDelegateCommand.

**Comment by dvorn** [comment]
@brianlagunas Just a quick thought:

Make DelegateCommand.Execute() return already-completed Task.
Make AsyncDelegateCommand inherit from DelegateCommand.
Make DelegateCommand.FromAsyncHandler return AsyncDelegateCommand instead of DelegateCommand.

**Comment by brianlagunas** [comment]
@dvorn I have to make changes to the DelegateCommandBase class.  I am playing around with removing the Task Execute, and just have the ICommand.Execute instead.  This will create parity between the way the command is invoked via code and a binding.  DelegateCommand redefines a number of methods.

**Comment by dvorn** [comment]
@brianlagunas It seems that sync and async cases are so different that DelegateCommandBase cannot have any code for Execute (except for delegating to derived classes).  But all CanExecute stuff may all be there.

**Comment by brianlagunas** [comment]
Check out #745 and let me know if this direction is looking good or not.  I'm not committed to this approach just yet, but it's a start.  And it will be a breaking change, which I am not 100% comfortable with, but it just wasn't done right the first time.

**Comment by dansiegel** [comment]
I would personally prefer to see this made in a way that leaves the static factory method intact for the next release. Make it obsolete of course, but at least make sure we all have time to update from ```DelegateCommand.FromAsyncHandler( ... )``` to what ever the new Async Delegate Command is.

**Comment by brianlagunas** [comment]
@dansiegel Yeah, this is just a real big pain in the ass. I'm not sure how I can make the necessary changes without breaking something.

**Comment by Emasoft** [comment]
I need to thank  @Robot1que and @dbeavon for rising this issue. I was considering using Prism for a big Xamarin project, but luckily you discovered this fatal flaw in its design. I really hope to see this fixed in the future. After being used to Prism, getting back to the old way of doing things will be painful!

**Comment by brianlagunas** [comment]
@Emasoft Just so you are aware, this is not a "fatal flaw".  This is the nature of async/await.  Commands by nature are not async, and do not support async/await.  No other framework even has this concept.  Also, exceptions being swallowed is only an issue if you are calling DelegateCommand.Execute manually in code without awaiting it and this is not an issue when data binding to the DelegateCommand.  This rally comes down to understanding how async/await works and how to properly handle exceptions in a multi-threaded application.

**Comment by dansiegel** [comment]
@Emasoft remember the Prism DelegateCommand is simply an implementation of Microsoft's ICommand which should be what your controls are referencing. Your controls are all going to be executing ```void Execute``` regardless of what you do. That said even in it's less than ideal state, the DelegateCommand works quite effectively across many applications WPF, UWP, and Xamarin Forms. If you feel the DelegateCommand from Prism is not the way to go right now, then I would suggest that you simply implement ICommand in any way you see fit or that you've seen someone else implement it and use that for your project. That shouldn't actually prevent you from using Prism.

**Comment by dvorn** [comment]
Created a PR #746 to resolve this issue keeping as much backward compatibility as possible.

**Comment by weitzhandler** [comment]
Hi everyone,
How about testing async command delegates?

    public class ViewModel
    {
      DelegateCommand _DoWorkCommand;
      public DelegateCommand DoWorkCommand => _DoWorkCommand ?? (_DoWorkCommand = new DelegateCommand(DoWork));
      async void DoWork()
      {
        await Task.Run(() => Thread.Sleep(5000));
        WorkDone = true;
      }

      public bool WorkDone { get; private set; }

      [Fact]
      public void TestDoWork()
      {
        DoWorkCommand.Execute();
        Assert.True(WorkDone);
      }
    }

The assertion fails because assertion happens before the task in `DoWork` has completed.

Am I wrong calling an `async void` from my command? Any nice general way to address this?

**Comment by brianlagunas** [comment]
I think you are over complicating things a bit.  Just make you method a Task and test it directly.

`public Task DoWork() { }`

You command can invoke it with:

`var command = new Command(async () => await DoWork());`

Now test your DoWork method directly.  Easy :)

**Comment by weitzhandler** [comment]
Sorry for not updating here, I've discovered that before, but since [this](https://github.com/PrismLibrary/Prism/issues/1002) issue, I switched to XF `Command`, and now I'm eating this $hit...
